### PR TITLE
Hackathon

### PR DESCRIPTION
**Pull request by silvester747**
Pull changes made during the TomTom Hackathon 2016:
- configurable aliases for add_library and add_executable
- configurable license comment
- allow and retain custom parts in cmake comments
- configure input files to ignore
- recursive generation
- refactoring, testing and bugfixes

### CODE REVIEW

**Comment by silvester747** [comment]
@dascandy please let me know how you want to go about getting this back to the origin. I was hoping to get this merged and then start working in my own fork.

File: config-cpp-dependencies.txt [hunk 1]
```diff
@@ -40,7 +55,44 @@ componentLocUpperLimit: 20000
...
+#addLibraryAlias: [
+#  add_my_library
+#  add_my_special_library
+#  ]
+
+# Aliases for CMake command add_executable. Each alias is assumed to take similar arguments as the
+# vanilla CMake command.
+# Each alias must be on its own line. The last line should only contain the closing bracket.
+#addExecutableAlias: [
+#  add_my_library
```
Comment (dascandy) [inline config-cpp-dependencies.txt]: That's not an executable.

File: src/CmakeRegen.cpp [hunk 2]
```diff
@@ -96,101 +115,31 @@ void RegenerateCmakeFilesForComponent(Component *comp, bool dryRun, bool writeTo
...
         if (writeToStdout) {
             std::cout << "######## End of " << comp->root.generic_string() << "/CMakeLists.txt\n";
             // No actual file written.
         } else if (dryRun) {
             if (FilesAreDifferent(comp->root / "CMakeLists.txt.generated", comp->root / "CMakeLists.txt")) {
                 std::cout << "Difference detected at " << comp->root << "\n";
+            } else {
+                filesystem::remove(comp->root / "CMakeLists.txt.generated");
             }
-            filesystem::remove(comp->root / "CMakeLists.txt.generated");
```
Comment (dascandy) [inline src/CmakeRegen.cpp]: Now a dry run will do something. Not sure I agree with this change...

File: src/main.cpp [hunk 3]
```diff
@@ -385,6 +402,10 @@ class Operations {
             std::cout << "\n";
         }
     }
+    void Recursive(std::vector<std::string>) {
+        recursive = true;
+        UnloadProject();
```
Comment (dascandy) [inline src/main.cpp]: Is this needed?

Comment (silvester747) [reply src/main.cpp]: I use it a lot in de TT version: only generate Engines/SomeEngine/... without touching the rest of the codebase.

File: config-cpp-dependencies.txt (hunk 1 above)
Comment (silvester747) [reply config-cpp-dependencies.txt]: Oops, will fix.

File: src/CmakeRegen.cpp (hunk 2 above)
Comment (silvester747) [reply src/CmakeRegen.cpp]: Will check why this was changed.

Comment (silvester747) [reply src/CmakeRegen.cpp]: I suspect this was a debugging leftover. Will change.

File: src/main.cpp (hunk 3 above)
Comment (dascandy) [reply src/main.cpp]: I meant the unloading of the project after applying this flag. Does it change the analysis result?

**Pull request merged**